metagraph/src/graph/alignment/aligner_chainer.cpp

```diff
@@ -0,0 +1,474 @@
+        return a.label_coordinates.size();
+    })) {
+        throw std::runtime_error("Post-chaining alignments with coordinates not supported");
+    }
+
+    DBGAlignerConfig no_chain_config { config };
+    no_chain_config.post_chain_alignments = false;
+    AlignmentAggregator<AlignmentCompare> aggregator(no_chain_config);
+
+    std::sort(alignments.begin(), alignments.end(), [](const auto &a, const auto &b) {
+        return std::make_tuple(a.get_orientation(),
+                               a.get_clipping() + a.get_query().size(),
+                               a.get_clipping(),
+                               b.get_score(),
+                               a.get_sequence().size())
+            < std::make_tuple(b.get_orientation(),
+                              b.get_clipping() + b.get_query().size(),
+                              b.get_clipping(),
+                              a.get_score(),
+                              b.get_sequence().size());
+    });
+
+    auto run = [&](bool rev_compl, auto begin, auto end) {
+        std::string_view this_query = rev_compl ? rc_query : query;
+        std::vector<score_t> best_score(this_query.size() + 1, 0);
+        for (auto it = begin; it != end; ++it) {
+            size_t end_pos = it->get_query().data() + it->get_query().size()
+                                - this_query.data();
+            if (it->get_score() > best_score[end_pos]) {
+                best_score[end_pos] = it->get_score();
+                construct_alignment_chain<AlignmentCompare>(
+                    node_overlap, config, this_query, Alignment(*it), it + 1, end,
+                    best_score, [&](Alignment&& chain) {
+                      aggregator.add_alignment(std::move(chain));
+                    }
+                );
+            }
+        }
+    };
+
+    // recursively construct chains
+    auto split_it = std::find_if(alignments.begin(), alignments.end(),
+                                 [](const auto &a) { return a.get_orientation(); });
+    run(false, alignments.begin(), split_it);
+    run(true, split_it, alignments.end());
+
+    return aggregator.get_alignments();
+}
+
+template <class AlignmentCompare>
+void construct_alignment_chain(size_t node_overlap,
+                               const DBGAlignerConfig &config,
+                               std::string_view query,
+                               Alignment&& chain,
+                               typename std::vector<Alignment>::iterator begin,
+                               typename std::vector<Alignment>::iterator end,
+                               std::vector<score_t> &best_score,
+                               const std::function<void(Alignment&&)> &callback) {
+    assert(begin <= end);
+    assert(chain.size());
+
+    const char *chain_begin = chain.get_query().data();
+    const char *chain_end = chain_begin + chain.get_query().size();
+    if (begin == end || chain_end == query.data() + query.size()) {
```

Comment:
What's the difference between `query` and `chain.get_query()`?

Reply:
I haven't gotten around to refactoring this yet, I want to rename `get_query` to `get_query_view`. `Alignment::get_query` returns the window in the query to which the path aligned. For precisely,
```c++
query.data() == chain.get_query().data() - chain.get_clipping();
query.data() + query.size() == chain.get_query().data() + chain.get_query().size() + chain.get_end_clipping()
```

Reply:
maybe `get_path_window()`, or something?

Reply:
if you have an idea, please go ahead and rename